### PR TITLE
[GSoC-Phase2] Add JSON Support

### PR DESCRIPTION
**Pull request by jiachengxu**
This PR contains the work of Phase 2 of my GSoC project #262:
- Implement the JSON Dataset support in C++ based on `jsoncpp`.
- Rewrite the test in `test_json.py` to make it more generic.
- Add a test of using `JSONDataset` with `tf.keras` based on [TensorFlow Custom training: walkthrough](https://www.tensorflow.org/tutorials/eager/custom_training_walkthrough).

Limition:
- In the current implementation, The `filename()` is called to get the filenames, and this will not work 
in file systems with scheme (e.g., `s3://`, `gcs://`, `azfs://` etc).

Next Step:
-  Get scheme file system (e..g, s3:// or gcs://) working.
- Add documentation/tutorial of JSON Dataset support.
- Implement `write_to_json` method.

/cc @yongtang @terrytangyuan

### CODE REVIEW

**Comment by yongtang** [comment]
@jiachengxu to help speed up your development, inside 
```c++
Status ReadRecord(io::InputStreamInterface* s, ..
```

you could call `filename()` to get the filename. Then you can try just process like a regular file with jsoncpp. 

The above will not work in file systems with scheme (e.g., `s3://`, `gcs://`, `azfs://` etc) but let's start getting basic workflow done before next step.

**Comment by jiachengxu** [comment]
Hi @yongtang, I have updated the PR with `ReadRecord` implemented, PTAL, and I am wondering if there is a way to only compile the JSON dataset source code for testing and debugging instead of compiling all the dataset support?

Thanks,
Jiacheng

File: tensorflow_io/json/kernels/json_input.cc [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ public:
+  Status ReadRecord(io::InputStreamInterface* s, IteratorContext* ctx, std::unique_ptr<io::BufferedInputStream>& state, int64 record_to_read, int64* record_read, std::vector<Tensor>* out_tensors) const override {
+    if (state.get() == nullptr) {
+      state.reset(new io::BufferedInputStream(s, 4096));
+    }
+
+    // Read the json file
+    ifstream ifs(filename());
+    Json::Reader reader;
+    Json::Value records;
```
Comment (yongtang) [inline tensorflow_io/json/kernels/json_input.cc]: @jiachengxu In this case, your `state` consists of ifstream, Json::Reader, and Json::Value, you should wrap them into one class, and use the newly defined class as part of the template.

`class JSONInput: public FileInput<io:BufferedInputStream> ` => `class JSONInput: public FileInput<YourNewClassWithJsonReaderValueAndIfstream> `

**Comment by yongtang** [comment]
@jiachengxu Also take a look at pandas' read_json with `records` type:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html

This will be the one we try to match.

File: tensorflow_io/json/kernels/json_input.cc (hunk 1 above)
Comment (jiachengxu) [reply tensorflow_io/json/kernels/json_input.cc]: Ok, I see, will update the PR now

**Comment by yongtang** [comment]
@jiachengxu Can you try to fix the test failures? Let me know if you have any questions. I am hoping this PR could be merged first. Then we can move to get scheme file system (e..g, `s3://` or `gcs://`) working in the follow up PR.

**Comment by jiachengxu** [comment]
PTAL, @yongtang @terrytangyuan, I wanted to request you guys for a review, but I don't think I have the permission of doing that :).

**Comment by jiachengxu** [comment]
Hi @yongtang, as we discussed in the weekly gsoc meeting today, a test of `JSONDataset` with `tf.keras` is necessary. 

I added a test of that in the latest commits. The dataset and the `tf.keras` model are quite similar to this [TensorFlow Custom training: walkthrough](https://www.tensorflow.org/tutorials/eager/custom_training_walkthrough) on the official TensorFlow website. I also use the same dataset but in JSON format for testing.

**Pull request merged**